src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,513 @@
+            {
+                if (key.Equals(entries[i].key))
+                    return true;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+
+            for (int i = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+                    (uint)i < (uint)entries.Length; i = entries[i].next)
+            {
+                if (key.Equals(entries[i].key))
+                    return entries[i].value;
+            }
+
+            return default;
+        }
+
+        public bool Remove(TKey key)
+        {
+            Entry[] entries = _entries;
+            int bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+            int entryIndex = GetEntryIndex(bucketIndex);
+
+            int lastIndex = -1;
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    if (lastIndex != -1)
+                    {   // Fixup preceding element in chain to point to next (if any)
+                        entries[lastIndex].next = entries[entryIndex].next;
+                    }
+                    else
+                    {   // Fixup bucket to new head (if any)
+                        _buckets[bucketIndex] = entries[entryIndex].next + 1;
+                    }
+
+                    entries[entryIndex] = default; // could use RuntimeHelpers.IsReferenceOrContainsReferences
+
+                    entries[entryIndex].next = -3 - _freeList; // New head of free list
+                    _freeList = entryIndex;
+
+                    Count--;
+                    return true;
+                }
+                lastIndex = entryIndex;
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                int bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+
+                for (int i = GetEntryIndex(bucketIndex);
+                        (uint)i < (uint)entries.Length; i = entries[i].next)
+                {
+                    if (key.Equals(entries[i].key))
+                        return ref entries[i].value;
+                }
+
+                return ref AddKey(key, bucketIndex);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private ref TValue AddKey(TKey key, int bucketIndex)
+        {
+            Entry[] entries = _entries;
+            int entryIndex;
+            if (_freeList != -1)
+            {
+                entryIndex = _freeList;
+                _freeList = -3 - entries[_freeList].next;
+            }
+            else
+            {
+                if (Count == entries.Length || entries.Length == 1)
```

Comment:
The null check that Dictionary does has to happen on reads. With this dummy array trick, it's only a check (for size 1) on writes.

Reply:
What's wrong with `Array.Empty`?

Reply:
division by zero on modulo

Reply:
Okay, that makes sense.

Reply:
After this goes in we should open an issue with suggestions for regular Dictionary. I have at least two: the dummy array trick to avoid the null check, and it seems it should be possible to remove their `freeCount` field - seems we proved it's not needed.

Reply:
Opened https://github.com/dotnet/corefx/issues/33392 for that.